core/shared/src/main/scala/fs2/async/mutable/Signal.scala

```diff
@@ -64,7 +62,15 @@ object Signal {
     new immutable.Signal[F, A] {
       def get = F.pure(a)
       def continuous = Stream.constant(a)
-      def discrete = Stream.empty // never changes, so never any updates
+
+      /**
+        * We put a single element here because otherwise the implementations of
+        * Signal as a Monad or Applicative get more annoying. In particular if
+        * this stream were empty, Applicatively zipping another Signal in the
+        * straightforward way would cause the (non-deterministically) zipped
+        * stream to be empty.
+        */
+      def discrete = Stream(a)
```

Comment:
This one is interesting; what do you mean by correct? Is there another `Signal` implementation I should look at? I see the linked pull request in another comment, but I'm not sure immediately how it relates. I'm assuming you're talking about invariants as they relate to `Signal`s themselves rather than say the Applicative laws (since Discipline is already checking those).

Presumably there are series of invariants/laws that we intuitively expect all `Signal`s to obey, although that's not spelled anywhere (perhaps they should be?).

My intuition is that there are the following invariants with some hand-waving around the word "equivalent" and "subsequence":

+ `continuous` should be equivalent to `Stream.repeatEval(get)`, even if it is not implemented as such.
+ `discrete` should be a subsequence of `continuous`; that is given a sufficiently fast consumer, all the elements observed in a `discrete` stream should be observed in a `continuous` stream and in the same order (although the `continuous` stream may have more elements).
+ `get`, `continuous`, and `discrete` should each be "side-effect-free;" that is even though there is an effect `F` parameter which indicates non-determinism, calling `get`, `continuous`, or `discrete` should not affect any other observers of the same signal (my `Signal` generator I use for the tests does *not* follow this because this makes testing more difficult).

I suspect you have another invariant in mind, in analogy to how the included implementations of `Topic` and `Queue` work:

+ Neither `continuous` nor `discrete` should ever terminate. Or perhaps this is a corollary of a deeper intuition that you have, that a downstream consumer should be unable to distinguish between a `Signal` that is "done" and a `Signal` which happens to be quiet for the moment, but could at any time receive a new change.

There's two things that are mildly annoying about this invariant. The first, superficial one is that it just makes testing much more difficult. Equivalence between `Signal`s probably has to be defined with some notion of convergence and without termination, I think your tests tend to become timeouts with all the crazy non-determinism around them. This is less true for `Applicative` where a sentinel value might still be enough (e.g. use `noneTerminate` with `async.hold` and wait on the `discrete` stream until it hits `None` and then check `get`). However, I strongly suspect that for the `Monad` instance (I do think `Signal` has a useful `Monad` instance, I'm just not sure how to write the `switchMap` that it depends on yet), you'll potentially have short-lived oscillation of values (e.g. with certain `flatMap` configurations you'll have a `Signal` that oscillates to `None` then away and then back again), which makes sentinel-based testing very difficult.

The somewhat less superficial one is that I think that behavior is exactly what I would expect of a `mutable.Signal`, but not necessarily of an `immutable.Signal`. I would lump `mutable.Signal` with all the other `mutable` structures, but I view an `immutable.Signal` as essentially just a `Stream` that happens to have a notion of "current" state which I can query at any time. And in the same way that a `Stream` can be infinite or finite, an `immutable.Signal` can be infinite or finite as well.

Reply:
Apologies for the wall of text :(

Reply:
Indeed like you observed, the problem with implementation that `def discrete = Stream(a)` is that that stream will terminate after `a` is produced, and that should not be the case. In fact discrete stream shall never ever terminate (same as continuous) unless explicitly told so (i.e. with Pipe). 

I am quite negative on having signals where one terminates discrete, while `continuos` is not terminates, and other variants of signals that works differently. 

I am not yet convinced by your arguments that we shall allow it. I am using signals for all tests in our code, and very, very rarely we need to rely on timeouts. Perhaps I am not understanding your point well. 
  


Also just few more remarks :

- `discrete` may not produce all elements available in `continuous`. There is no queue-like guarantee. 
- the only difference between `discrete` and `continuous` is that `discrete` will _hold_ if there is no new value signalled, unlike `continuous` that always produces value. 
- I don't think so there shall be any difference in behaviour of mutable/immutable signal. If we would allow it, then I think the `mutable` Signal must not extends `immutable` signal variant. 
- I believe that current implementation of `mutable.Signal.constant#discrete` is incorrect (buggy), and it should in fact be Stream(a) ++ eval(F.never).

Reply:
On reflection I'm more torn than I used to be; I could be swayed that the behavior should be non-terminating. I still think though, given that this behavior has existed since 0.9, for the sake of backwards-compatibility that's a change that should hold off until 1.0.

Reply:
@changlinli essentially current implementation broke following law : 

````
async.signalOf[A](a).map(identity) == Signal.constant(a)
```
so I strongly believe even if it is now in 0.10.x is incorrectly defined, it shall be fixed to 1.0  release. I think it can also be fixed in 0.10 tree, perhaps with remark in readme.

Reply:
@changlinli I think lets focus this to 1.0 release. I would only suggest to make a patch to 0.10 to fix that constant signal constructor, if you will have some spare time by separate PR. 
Thank you so much for this work!.

Reply:
I'm fine with retargeting this at 1.0 branch.

Reply:
I think 0.10 is ok with changing constant's discrete definition.  Not sure about binary compatibility though. I think if it breaks, than this is red light to go to 0.10. Can you show where the binary compatibility is broken?

Reply:
In order to never terminate, you need `IO.never`, which you can access through an `Effect[F].liftIO`. This changes the signature of `constant`.

```scala
// Old signature
def constant[F[_], A](a: A)(implicit F: Applicative[F]): immutable.Signal[F, A]
// New signature
def constant[F[_], A](a: A)(implicit F: Effect[F]): immutable.Signal[F, A]
```

Appropriately `mimaReportBinaryIssues` reports back

```
[info] fs2-core: found 1 potential binary incompatibilities while checking against co.fs2:fs2-core_2.12:0.10.4 
[error]  * method constant(java.lang.Object,cats.Applicative)fs2.async.immutable.Signal in object fs2.async.mutable.Signal's type is different in current version, where it is (java.lang.Object,cats.effect.Effect)fs2.async.immutable.Signal instead of (java.lang.Object,cats.Applicative)fs2.async.immutable.Signal
[error]    filter with: ProblemFilters.exclude[IncompatibleMethTypeProblem]("fs2.async.mutable.Signal.constant")
```

Reply:
Only that we can't make that signature change in 0.10

Reply:
@mpilquist @pchlupacek So can I push again for the `Applicative` instance in 0.10 and then hold off on the non-termination until 1.0? It passes `Applicative` laws either way (although it's a lot harder to show if you put non-termination in)

Reply:
That sounds good to me.

Reply:
Cool, I'm still blocked on your review; is there anything else you'd like me to fix up?